### PR TITLE
Add named volume for mounting host files via OLYMPIA_MOUNT

### PR DESCRIPTION
**Pull request by KevinMind**
Rleates to: mozilla/addons#15066

### Description

- adds OLYMPIA_MOUNT/OLYMPIA_UID setting to inform if host files are mounted
- merge docker-compose.ci.yml as we now control host mounting via the argument above
- add test checking host file presence in the container

### Context

Follow up to https://github.com/mozilla/addons-server/pull/22912

By introducing a specific argument for the desired mounting strategy we can unify our docker compose files and have a simpler flow for mounting files correctly.

We must use bind mounts when mounting files from the host, named volumes (even with a host mounted device) do not send appropriate inotify events in a timely manner. This sucks and should be fixed in docker but for now it seems safer to use bind mounts

We map arguments `OLYMPIA_*` to `HOST_*` in the .env file. This is because the `setup.py` script does not always accept the user input and we don't want to let an internal command (like docker compose `<whatever>`) to rely on a potentially invalid value lingering on the environment but instead only read the explicitly set values in the .env. By using different names in that transaction we can guarantee this.

It is possible that we should raise errors in setup.py when invalid values are passed for OLYMPIA_MOUNT but this patch introduces enough changes and we can see if this is an issue and raise it then.

### Prod Mode

To verify expected mount behavior, run each:

```bash
make up DOCKER_TARGET=production OLYMPIA_MOUNT=<production|development>
```

- expect the .env to have a value `HOST_MOUNT` equal to the specified value
- expect the web/worker containers to have an environment variable set to the specified value
- expect settings.OLYMPIA_MOUNT to equal the specified value

- If `developmemnt` is provided
    - Expect the .env to have a value `HOST_MOUNT_SOURCE` equal to `"./"`
    - expect docker compose to use a bind mount for the /data/olympia volume
- if `production` is provided
    - Expect the .env to have a value `HOST_MOUNT_SOURCE` equal to `"data_olympia_"`
    - expect docker compose to use a named volume `data_olympia` that is not mounted on the host. (changing files on the host should not impact files in the container, no uwsgi reload)

To test this, attach a shell in the containers and inspect the files. Additionally, you can run the check command to verify expectations.

Additionally, verify that volumes are correctly mounted to `nginx`

- /data/olympia -> the same files in web/worker
- /data/olympia/storage -> files in the shared storage bind mount
- /data/olympia/site-static -> files in the web container 

And to `worker`
- /data/olympia -> the same files in web/worker
- /data/olympia/storage -> files in the shared storage bind mount

> [!IMPORTANT]
> Worker gets shared storage but not static files. these should be accessed via URL in the worker container if at all

### Dev Mode

When running the project in dev mode (DOCKER_TARGET=dev) then the OLYMPIA_MOUNT will always be set to dev, even if specified otherwise.

```bash
make up DOCKER_TARGET=production OLYMPIA_MOUNT=<production|development>
```

- Run the assertions in Prod mode, but always assuming that OLYMPIA_MOUNT is "development" regardless of docker target

### Checklist

- [X] Add `#ISSUENUM` at the top of your PR to an existing open issue in the mozilla/addons repository.
- [X] Successfully verified the change locally.
- [X] The change is covered by automated tests, or otherwise indicated why doing so is unnecessary/impossible.
- [ ] Add before and after screenshots (Only for changes that impact the UI).
- [x] Add or update relevant [docs](../docs/) reflecting the changes made.

### CODE REVIEW

**Comment by diox** [comment]
Could you resolve the few conflicts in that PR ?

**Comment by KevinMind** [comment]
@diox this is now done, and:

> Could you resolve the few conflicts in that PR ?

Turns out this PR was too primitive to land after some of the other PRs landed. The goal was to strip as much out as possible to make it smaller. That doesn't work anymore so I added as little back until it is now working.

- merged the CI compose file
- use mapping of OLYMPIA_MOUNT to make the container behave correctly.

**Review by eviljeff — COMMENTED** [review]
🤔

File: docker-compose.yml [hunk 1]
```diff
@@ -20,10 +20,20 @@ x-env-mapping: &env
...
     - CIRCLECI
-    - HOST_UID
-    - DEBUG
     - DATA_BACKUP_SKIP
 
+# Value is determined by the setup.py script
+# production: (data_olympia):/data/olympia/storage
+# development: (./):/data/olympia/storage
+x-olympia-mount: &olympia-mount
+  ${HOST_MOUNT_SOURCE:?}:/data/olympia
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply docker-compose.yml:29]: We should not. This syntax `:?` is actually requiring this value to be set. This value must be defined and a valid value and there are no circumstances where it should be undefined.

File: docker/nginx/addons.conf [hunk 2]
```diff
@@ -7,11 +7,11 @@ server {
 
     location /data/olympia/storage/ {
         internal;
-        alias /srv/storage/;
+        alias /data/olympia/storage/;
```
[the remark this answers is not in the document]
Comment (KevinMind) [reply docker/nginx/addons.conf:10]: `/srv` is a totally arbitrary path defined in our docker compose file. We can make it anything that we want. Using `/data/olympia` allows us to use one definition of the `*storage-mount` in the compose file which simplifies the code.

We _could_ keep it as `/srv` but then we couldn't "abstract" the mount definition. I'm happy to do that if you feel strongly it's not a need to have.

But, FWIW, changing this here has zero impact on any production environments as our storage path in django is defined via settings/environment variables and it only matters that the actual path matches the expected one, which this code satisfies.

[the remark this answers is not in the document]
Comment (eviljeff) [reply docker/nginx/addons.conf:10]: Can we keep it as `/srv` for now then?  I'm uncomfortable approving that change without further verification with various repos and people, so if it's an option to not change it then it doesn't block the approval.

[the remark this answers is not in the document]
Comment (KevinMind) [reply docker/nginx/addons.conf:10]: Sure thing.

[the remark this answers is not in the document]
Comment (KevinMind) [reply docker/nginx/addons.conf]: Makes the docker-compose a bit uglier, but we can fix this in a follow up.

[the remark this answers is not in the document]
Comment (KevinMind) [reply docker/nginx/addons.conf]: FWIW the root value for storage is defined [here](https://github.com/search?q=repo%3Amozilla-it%2Fwebservices-infra%20%22NETAPP_STORAGE_ROOT%22&type=code)

And that value is set in production [here](https://github.com/search?q=repo%3Amozilla-it%2Fwebservices-infra%20%22NETAPP_STORAGE_ROOT%22&type=code)

In production environments (I don't know the details) there is a mechanism to sync this file path between web and worker and IIRC we serve those files via URL using [nginx](https://github.com/mozilla-it/webservices-infra/blob/3222c6c6151ab0af2bb8607d0629e0c537d4636f/amo/k8s/amo-proxy/configs/addons.conf.tpl)

..Which [itself](https://github.com/mozilla-it/webservices-infra/blob/3222c6c6151ab0af2bb8607d0629e0c537d4636f/amo/k8s/amo-proxy/values.yaml#L102) seems to point back to the same mount path that the NET_APP_STORAGE variable points to.

[the remark this answers is not in the document]
Comment (KevinMind) [reply docker/nginx/addons.conf]: I believe (under the hood) we are using some blob storage in GCP which is mounted on that path, effectively allowing nginx/web/worker to use the same storage file system even though the actual files are in lala land.

[the remark this answers is not in the document]
Comment (KevinMind) [reply docker/nginx/addons.conf]: It's kind of exactly what is happening here, we just use a different path. The actual name of the path is almost completely irrelevant.

@diox or @bqbn can tell me I'm completely wrong on this.. I hope I'm not.

**Review by eviljeff — APPROVED** [review]

**Pull request merged**